ouroboros-consensus/test/infra-test/Ouroboros/Network/Driver/Limits/Extras.hs

```diff
@@ -0,0 +1,93 @@
+{-# LANGUAGE QuantifiedConstraints #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TypeFamilies          #-}
+
+module Ouroboros.Network.Driver.Limits.Extras (module Ouroboros.Network.Driver.Limits.Extras) where
+
+import           Cardano.Slotting.Time (SlotLength, getSlotLength)
+import           Control.Monad.Class.MonadTimer.SI (MonadTimer)
+import           Control.Tracer (Contravariant (contramap), Tracer)
+import           Data.Time.Clock (secondsToDiffTime)
+import qualified Network.TypedProtocol as TP
+import qualified Network.TypedProtocol.Codec as TP
+import           Ouroboros.Consensus.Util (ShowProxy)
+import           Ouroboros.Consensus.Util.IOLike (DiffTime,
+                     MonadAsync (concurrently), MonadFork, MonadMask,
+                     MonadSTM (STM), MonadThrow)
+import           Ouroboros.Network.Channel (Channel)
+import           Ouroboros.Network.Driver (TraceSendRecv, runPeer)
+import           Ouroboros.Network.Driver.Limits (runPipelinedPeerWithLimits)
+import           Ouroboros.Network.Driver.Simple (Role (Client, Server))
+import           Ouroboros.Network.Protocol.ChainSync.Codec
+                     (ChainSyncTimeout (..), byteLimitsChainSync)
+import           Ouroboros.Network.Protocol.ChainSync.Type (ChainSync,
+                     ClientHasAgency, ServerHasAgency)
+import           Ouroboros.Network.Protocol.Limits (ProtocolSizeLimits (..),
+                     ProtocolTimeLimits (..), shortWait)
+import           Test.Ouroboros.Consensus.ChainGenerator.Params (Asc, ascVal)
+import           Test.Util.Orphans.IOLike ()
+
+-- | Same as 'runConnectedPeersPipelined' except the client peer is ran not with
+-- 'runPipelinedPeer' but with 'runPipelinedPeerWithLimits'.
+runConnectedPeersPipelinedWithLimits ::
+  ( MonadAsync m
+  , MonadFork m
+  , MonadMask m
+  , MonadThrow (STM m)
+  , MonadTimer m
+  , Show failure
+  , forall (st' :: ps). Show (ClientHasAgency st')
+  , forall (st' :: ps). Show (ServerHasAgency st')
+  , ShowProxy ps
+  ) =>
+  m (Channel m bytes, Channel m bytes) ->
+  Tracer m (Role, TraceSendRecv ps) ->
+  TP.Codec ps failure m bytes ->
+  ProtocolSizeLimits ps bytes ->
+  ProtocolTimeLimits ps ->
+  TP.PeerPipelined ps pr st m a ->
+  TP.Peer ps (TP.FlipAgency pr) st m b ->
+  m (a, b)
+runConnectedPeersPipelinedWithLimits createChannels tracer codec sizeLimits timeLimits client server =
+    createChannels >>= \(clientChannel, serverChannel) ->
+    (fst <$> runPipelinedPeerWithLimits tracerClient codec sizeLimits timeLimits clientChannel client)
+      `concurrently`
+    (fst <$> runPeer tracerServer codec serverChannel server)
+  where
+    tracerClient = contramap ((,) Client) tracer
+    tracerServer = contramap ((,) Server) tracer
+
+chainSyncNoSizeLimits :: ProtocolSizeLimits (ChainSync header point tip) bytes
+chainSyncNoSizeLimits = byteLimitsChainSync (const 0)
+
+chainSyncTimeouts ::
```

Comment:
In a call today with Torsten and Alex we concluded that it would be fine to make the mustReply timeout almost as short as the others. The long timeout is only needed when the client is caught up and the server is just waiting for another block to be minted.

Reply:
> It would be nice to have some indication of how the timeouts depend on the parameters in this function.

More than the comment on `mustReplyTimeout`?

> Related to the above, I don't understand yet how we could mess up if we decide the timeouts ignoring `SlotLength` and `Asc`.

`SlotLength` and `Asc` together give you the average time that you have to wait between two blocks being minted. `SlotLength` is a constant in our tests (20) but `Asc` is picked randomly by the schema generator. Picking timeouts that ignore those parameters might lead to uselessly long timeouts (and then we might as well have no timeouts at all) or way too short timeouts (making the node under test disconnect from even an alert node).

> Also, the fields of `ChainSyncTimeout` are undocumented at its definition. Have you earned any insights from working with this code?

Each field is a `Maybe DiffTime` describing the absence (with `Nothing`) or the presence (with `Just dt`) of a timeout. The names of the field match that of the states in the `ChainSync` mini-protocol where the _other peer_ (here, the server) has agency. I suppose we could add a comment stating something like that near the `ChainSyncTimeout` definition.

> In a call today with Torsten and Alex we concluded that it would be fine to make the mustReply timeout almost as short as the others. The long timeout is only needed when the client is caught up and the server is just waiting for another block to be minted.

The current implementation of the scheduled server never sends `AwaitReply` so we never end up in the state where the `mustReply` timeout matters. It will indeed only matter when caught up. I don't understand why we would make it shorter. Either it doesn't matter because we're not caught up so never in that state or it does because we're caught up and our tests might include that sometimes and then even an alert peer will get disconnected from.

Reply:
> Either it doesn't matter because we're not caught up so never in that state or it does because we're caught up and our tests might include that sometimes and then even an alert peer will get disconnected from.

The node under test might know that it is not caught up, but the server might send a MsgAwaitReply message anyway (dishonestly if it is not well connected or adversarial). It wouldn't make sense for the node under test to wait long in this case. The computation constructing the timeout could query the node to see if it is caught up or not, and set the timeout accordingly.

> More than the comment on mustReplyTimeout?

I think the comment fails to explain why we would care about the probabilities of empty slot streaks. The discussion so far has made it clear that it is relevant when the node under test is caught up, because then the server needs to wait for a new block to be minted (this is the bit I missed). But moreover, the node under test is not always caught up.

Thanks for your comments. They brought me clarity.

Reply:
> The node under test might know that it is not caught up, but the server might send a MsgAwaitReply message anyway (dishonestly if it is not well connected or adversarial). It wouldn't make sense for the node under test to wait long in this case. The computation constructing the timeout could query the node to see if it is caught up or not, and set the timeout accordingly.

I don't think that is the right way to look at it. In my opinion, it is a perfectly acceptable scenario for the node under test to not be caught-up and for an adversary to dishonestly send a `MsgAwaitReply`. It could happen in a real world setting as well; a form of leashing attack, I suppose. If this gets the node under test stalled or blocked, that is a bug, and we do want the tests to exhibit it. I think the timeouts should be set as reasonable value that a real node would use (which is what I tried to do so far) for a chain as generated by the `ChainGenerator`.

> I think the comment fails to explain why we would care about the probabilities of empty slot streaks. The discussion so far has made it clear that it is relevant when the node under test is caught up, because then the server needs to wait for a new block to be minted (this is the bit I missed). But moreover, the node under test is not always caught up.

Fair enough. I am not sure that the right place is near these timeouts definitions but it will not hurt either. I will suggest a minuscule PR with this and we can talk about it there.

Reply:
> I think the timeouts should be set as reasonable value that a real node would use (which is what I tried to do so far) for a chain as generated by the ChainGenerator.

That sounds right to me. But isn't the timeout set by the Genesis node depending on its state (caught up or not)?

A long timeout wouldn't make sense in the synchronizing state. It would mean maintaining a connection to a peer which is not well connected.